runtime/caml/eventring.h

```diff
@@ -0,0 +1,303 @@
+    EV_MAJOR_GC_CYCLE_DOMAINS,
+    EV_MAJOR_GC_PHASE_CHANGE,
+    EV_MAJOR_GC_STW,
+    EV_MAJOR_MARK_OPPORTUNISTIC,
+    EV_MAJOR_SLICE,
+    EV_MAJOR_FINISH_CYCLE,
+    EV_MINOR_CLEAR,
+    EV_MINOR_FINALIZERS_OLDIFY,
+    EV_MINOR_GLOBAL_ROOTS,
+    EV_MINOR_LEAVE_BARRIER,
+    EV_STW_API_BARRIER,
+    EV_STW_HANDLER,
+    EV_STW_LEADER,
+    EV_MAJOR_FINISH_SWEEPING,
+    EV_MINOR_FINALIZERS_ADMIN,
+    EV_MINOR_REMEMBERED_SET,
+    EV_MINOR_REMEMBERED_SET_PROMOTE,
+    EV_MINOR_LOCAL_ROOTS_PROMOTE,
+    EV_DOMAIN_CONDITION_WAIT
+} ev_runtime_phase;
+
+typedef enum {
+    EV_C_ALLOC_JUMP,
+    EV_C_FORCE_MINOR_ALLOC_SMALL,
+    EV_C_FORCE_MINOR_MAKE_VECT,
+    EV_C_FORCE_MINOR_SET_MINOR_HEAP_SIZE,
+    EV_C_FORCE_MINOR_WEAK,
+    EV_C_FORCE_MINOR_MEMPROF,
+    EV_C_MAJOR_MARK_SLICE_REMAIN,
+    EV_C_MAJOR_MARK_SLICE_FIELDS,
+    EV_C_MAJOR_MARK_SLICE_POINTERS,
+    EV_C_MAJOR_WORK_EXTRA,
+    EV_C_MAJOR_WORK_MARK,
+    EV_C_MAJOR_WORK_SWEEP,
+    EV_C_MINOR_PROMOTED,
+    EV_C_MINOR_ALLOCATED,
+    EV_C_REQUEST_MAJOR_ALLOC_SHR,
+    EV_C_REQUEST_MAJOR_ADJUST_GC_SPEED,
+    EV_C_REQUEST_MINOR_REALLOC_REF_TABLE,
+    EV_C_REQUEST_MINOR_REALLOC_EPHE_REF_TABLE,
+    EV_C_REQUEST_MINOR_REALLOC_CUSTOM_TABLE
+} ev_runtime_counter;
+
+/* external C-API for reading from the eventring */
+struct caml_eventring_cursor;
+
+typedef enum {
+  E_SUCCESS = 0,
+  E_CURSOR_NOT_OPEN = -1,
+  E_CORRUPT_STREAM = -2,
+  E_ALLOC_FAIL = -3,
+  E_PATH_FAILURE = -4,
+  E_OPEN_FAILURE = -5,
+  E_NO_CURRENT_RING = -6,
+  E_MAP_FAILURE = -7,
+} eventring_error;
+
+/* Starts eventring. Needs to be called before [caml_eventring_create_cursor] */
+extern value caml_eventring_start();
+
+/* Pauses eventring. No new events (other than the pause itself) will be written
+   to the eventrings by this domain immediately and all other domains soon. */
+extern value caml_eventring_pause();
+
+/* Removes eventring. New events (as well as a resume event) will be written to
+   this domain immediately and all other domains soon. */
+extern value caml_eventring_resume();
+
+/* Create a cursor to read events from an eventring. Cursors can be created for
+   eventrings in and out of process. To create one for the current process, pass
+   [eventring_path] as NULL and a [pid] < 0. Otherwise [eventring_path] is a
+   path to a directory containing the .eventring files. [pid] is the process id
+   (or equivalent) of the startup OCaml process. The resulting cursor can be
+   used with `caml_eventring_read_poll` to read events from the eventrings. */
+extern eventring_error
+caml_eventring_create_cursor(const char_os* eventring_path, int pid,
+                             struct caml_eventring_cursor **cursor_res);
+
+/* Set the runtime_begin event callback on the cursor */
+extern void caml_eventring_set_runtime_begin(
+    struct caml_eventring_cursor *cursor,
+    int (*f)(int domain_id, void *callback_data, uint64_t timestamp,
+             ev_runtime_phase phase));
```

Comment:
Naive questions:
- Is there a way to un-set callbacks, and does that provide performance benefits afterwards?
- How does it work if several libraries want to observe events, won't they destroy each other `set` calls? Is the idea that they each independently create their own cursor? (Could that be suggested somewhere in the API documentation?)

Reply:
So yes, the intention is that each consumer would have their own cursor and can choose the set of callbacks they wanted to receive. I've added a couple of sentences around that to the documentation for `caml_eventring_create_cursor` in https://github.com/ocaml/ocaml/pull/10964/commits/a539ba44024a97c9d5ce336f8b7f4c45a8176839 and in `Eventring.mli` in https://github.com/ocaml/ocaml/pull/10964/commits/9d4403810c28d3a717c6953c172d007edc50b3a0